### PR TITLE
Refactor DeviceStateHolder and its receivers into RunConditionMonitor

### PR DESCRIPTION
**Pull request by ghost**
Purpose
Run conditions were detected async by listening to "incremental" information retrieved by broadcast receivers and "statically" by reading current device state or config. With that mixture, a "memory" of the current run/run-not state cannot be held plus "race" timing issues occur if the user changes preferences resulting in the "syncthing is disabled" dialog when it should run according to the newly set options.

The refactored code does exactly the same as the original code. From my point of view, the new implementation of the RunConditionMonitor fits better for future expansion and reflects the need of re-thinking the run condition preferences. I already prepared functions to properly detect mobile data and flight mode which will then be used in a follow-up PR.

As a plus, we'll also be able to make a nicer screen explaining the user "why syncthing is disabled" in MainActivity as all asynchronous information collection that previously would have made it hard to determine which circumstance was the cause for enabling/disabling syncthing was refactored into synchronous linear condition checks. To save phone's battery, the check of which conditions apply is still triggered by broadcasts received from the android system.

When completed, this will be the base to fix the issues:
- #1173 "Disable Syncthing in Flightmode"
- #937 "Add run condition: Avoid metered networks"
- #729 "Show reason why Syncthing is stopped"
- #588 "Option to link Syncthing to Global Android "Auto-sync Data""
- #549 ""WiFi only" should not depend on "background/foreground", Wifi only should be default"

Testing
Verified working on device lg-h815 running Android 7.1.2 at commithttps://github.com/syncthing/syncthing-android/pull/1187/commits/b2ea83d47b4ae2a01f5ce96b5fd76a1215f35844  .

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+            return false;
+        }
+        if (!ni.isConnected()) {
+            // No network connection.
+            return false;
+        }
+        return cm.isActiveNetworkMetered();
+    }
+
+    private boolean isMobileDataConnection() {
```
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java]: Not used?

**Review by AudriusButkevicius — COMMENTED** [review]
So this looks all good to me apart from a few unused functions, which I am not sure I understand the purpose off.

File: app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java (hunk 1 above)
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java]: Currently yes. Is it okay if I'll need that function later as I intend to make a new rjn condition gui proposal in a follow-up pr?

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java]: I think that's when you should add these functions, in a PR where they are actually used.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java]: Ok, removed them. Ready?

**Review by AudriusButkevicius — APPROVED** [review]

**Comment by AudriusButkevicius** [comment]
If you plan some features shortly, let me know, and I'll hold back the release.

**Comment by Catfriend1** [comment]
Thanks for leaving a note on this. From my side, it's okay to make a release. The next enhancements I've planned regarding the run conditions aren't urgent so they can wait to be released :).

**Comment by Catfriend1** [comment]
I guess the release will contain 14.49? Then everything is okay as I already did my testing on 14.49-rc.4 with the current master. Thanks for your work on "syncthing core", thumbs up for 14.49 to be finally there :)

**Pull request merged**